federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,230 @@ limitations under the License.
 
 package service
 
+import (
+	"fmt"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+)
+
+const (
+	// minDnsTtl is the minimum safe DNS TTL value to use (in seconds).  Less than this, and some caching DNS servers might ignore it.
+	minDnsTtl = 180
+)
+
+// getHealthyEndpoints returns the hostnames and/or IP addresses of healthy endpoints for the service, at a zone, region and global level (or an error)
+func (s *ServiceController) getHealthyEndpoints(clusterName string, cachedService *cachedService) (zoneEndpoints, regionEndpoints, globalEndpoints []string, err error) {
+	zoneEndpoints = []string{}
+	regionEndpoints = []string{}
+	globalEndpoints = []string{}
+	zoneName, err := s.getClusterZoneName(clusterName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	regionName, err := s.getClusterRegionName(clusterName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	for lbClusterName, lbStatus := range cachedService.serviceStatusMap {
+		lbZoneName, err := s.getClusterZoneName(lbClusterName)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		lbRegionName, err := s.getClusterRegionName(lbClusterName)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		for _, ingress := range lbStatus.Ingress {
+			var address string
+			// We should get either an IP address or a hostname - use whichever one we get
+			if ingress.IP != "" {
+				address = ingress.IP
+			}
+			if ingress.Hostname != "" {
+				address = ingress.Hostname
+			}
+			if len(address) <= 0 {
+				return nil, nil, nil, fmt.Errorf(
+					"Service %s/%s in cluster %s has neither LoadBalancerStatus.ingress.ip nor LoadBalancerStatus.ingress.hostname."+
+						"Cannot use it as endpoint for federated service.",
```

Comment:
Make this a single line. Go has a general convention of not breaking lines this way. Also, the indentation is weird now anyway.